src/circuit_builder.rs

```diff
@@ -0,0 +1,166 @@
+};
+
+use crate::circuit::{Gate, GateAddr, GateOp, GeneralCircuit, Layer};
+
+#[derive(Debug, Clone)]
+struct Builder {
+    // contains a vec of all Layers
+    layers: Vec<Vec<Node>>,
+}
+
+#[derive(Debug, Clone)]
+struct Node {
+    // This is the operation of the gate
+    // TODO: make this a reference to an operation
+    op: Option<GateOp>,
+    // Used for checking if the node has been processed
+    is_processed: bool,
+    // The left child of the node
+    left_child: Option<GateAddr>,
+    // The right side of the node
+    right_child: Option<GateAddr>,
+    // The layer index and gate index of the node
+    node_addr: Option<GateAddr>,
+}
+
+impl Node {
+    pub fn new(
+        op: Option<GateOp>,
+        left_child: Option<GateAddr>,
+        right_child: Option<GateAddr>,
+        node_addr: Option<GateAddr>,
+    ) -> Node {
+        Self {
+            op,
+            is_processed: false,
+            left_child,
+            right_child,
+            node_addr,
+        }
+    }
+
+    fn is_input(&self) -> bool {
+        self.left_child.is_none() && self.right_child.is_none() && self.op.is_none()
+    }
+
+    fn to_gate(&self) -> Gate {
+        // Create gate with left and right children
+        let inputs = [self.left_child.unwrap(), self.right_child.unwrap()];
+        Gate::new(self.op.clone().expect("Operation should exist"), inputs)
+    }
+}
+
+impl Builder {
+    // Initializes the builder
+    fn init() -> Self {
+        Self { layers: vec![] }
+    }
+
+    // Creates an input node
+    fn create_input_node(&mut self) -> GateAddr {
+        // Inputs are on layer 0
+        if self.layers.len() <= 0 {
+            self.layers.push(vec![]);
+        }
+
+        let input_layer = &mut self.layers[0];
+
+        let gate_index = input_layer.len();
+
+        let node = Node::new(None, None, None, Some((0, gate_index)));
+
+        input_layer.push(node);
+
+        (0, gate_index)
+    }
+
+    // Adds a gate to the circuit and returns the gate
+    fn add_gate(&mut self, left_child: GateAddr, right_child: GateAddr, op: &GateOp) -> GateAddr {
+        let node_layer = max(left_child.0, right_child.0) + 1;
+
+        if self.layers.len() <= node_layer {
+            self.layers.push(vec![]);
+        }
+
+        let layer = &mut self.layers[node_layer];
+
+        let node_index = layer.len();
+
+        let node = Node::new(
+            Some(op.clone()),
+            Some(left_child),
+            Some(right_child),
+            Some((node_layer, node_index)),
+        );
+
+        layer.push(node);
+
+        (node_layer, node_index)
+    }
+
+    fn get_node_mut() {
+        todo!()
+    }
+
+    // Builds the layered circuit
+    fn build_circuit(&mut self) -> GeneralCircuit {
+        // Skips the input layer
+        let layers = self
+            .layers
+            .iter()
+            .skip(1)
+            .map(|layer| Layer::new(layer.iter().map(|node| node.to_gate()).collect()))
```

Comment:
It might be possible to pass in the total layer count to node.to_gate(layer_count), then have that flip the indexes when creating the GeneralCircuit gates.